File: second/array/rotate-array.c
void reverse(int *nums, int start, int end)
{
    int tmp;
    while(start < end){
        tmp = nums[start];
        nums[start] = nums[end];
        nums[end] = tmp;

        ++start;
        --end;
    }
}

void rotate(int* nums, int numsSize, int k) {
    k %= numsSize;
    reverse(nums,0,numsSize - k - 1);
    reverse(nums,numsSize - k,numsSize -1);
    reverse(nums,0,numsSize-1);
}

File: go/rotate-array.go
func rotate(nums []int, k int)  {
    k = k % len(nums)

    for k > 0 {
        var tmp = nums[len(nums) - 1]
        for i := len(nums) - 1; i > 0; i-- {
            nums[i] = nums[i - 1]
        }
        nums[0] = tmp
        k--
    }
}
